Lib/typing.py

```diff
@@ -230,15 +230,11 @@ def _remove_dups_flatten(parameters):
         else:
             params.append(p)
     # Weed out strict duplicates, preserving the first of each occurrence.
-    all_params = set(params)
-    if len(all_params) < len(params):
-        new_params = []
-        for t in params:
-            if t in all_params:
-                new_params.append(t)
-                all_params.remove(t)
-        params = new_params
-        assert not all_params, all_params
+    new_params = []
+    for p in params:
+        if p not in new_params:
+            new_params.append(p)
+    params = new_params
     return tuple(params)
```

Comment:
This whole bit can be sped up by just replacing it with:

```py
return tuple(dict.fromkeys(params))
```

Reply:
Thanks, I will add some tests. By the way, `params` is a list, `tuple(dict.fromkeys(params))` seems not faster than `tuple(params)`, I think it would be better to keep `tuple(params)`:

![20190901004654](https://user-images.githubusercontent.com/7546325/64066889-084a9180-cc52-11e9-992b-2895e4a4474e.png)

Reply:
Sorry, I should have been clearer! I meant replacing this whole list + loop bit with that one line. `dict.fromkeys` is a nice shortcut to remove duplicates while preserving order.

Reply:
Ah... I see now. You're trying to avoid the (broken) hashing that the competing PR #15400 fixes.

I don't know how I feel about that. I think hashing should be fixed rather than just sidestepped here, since it's so tied to equality. At this point I'm actually in favor of the approach the other PR takes.

@gvanrossum and @ilevkivskyi, thoughts?

Reply:
@hongweipeng you can revert your most recent commit (3676a50, “code review”). My comment was misguided!